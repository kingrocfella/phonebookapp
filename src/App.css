.display-end {
  display: flex;
  justify-content: flex-end;
}

.user-avatar {
  background-color: blue;
  border-radius: 50%;
  width: 3rem;
  height: 3rem;
  color: white;
}

.cursor {
  cursor: pointer;
}
