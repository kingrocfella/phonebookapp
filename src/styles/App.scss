@import "./variables";

html,
body {
  background-color: $secondary-color;
}

.display-end {
  display: flex;
  justify-content: flex-end;
}

.display-center-end {
  display: flex;
  justify-content: flex-end;
  align-items: center;
}

.primary-button {
  background-color: $primary-color;
  color: white;
  border-radius: 1.5rem;
  padding: $btnpadding;
}

.block-button {
  background-color: $primary-color;
  color: white;
}

.delete-button {
  background-color: $danger;
  color: white;
  border-radius: 1.5rem;
  padding: $btnpadding;
}

.user-avatar {
  background-color: $primary-color;
  border-radius: 50%;
  width: 3rem;
  height: 3rem;
  color: white;
}

.cursor {
  cursor: pointer;
}

.display-center {
  display: flex;
  justify-content: center;
  align-items: center;
}

.display-vertical-center {
  display: flex;
  align-items: center;
}

.hovered-row {
  background-color: $cardboxshadow;
}

.contact-table {
  border-spacing: 0rem;
  border-collapse: separate;
}

.space-between {
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.invisible {
  visibility: hidden;
}
