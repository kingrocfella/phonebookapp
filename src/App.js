import { useState } from "react";
import Modal from "./components/Modal";
import AddContact from "./components/addContact";
import ContactList from "./components/contactList";
import ContactDetails from "./components/contactDetails";
import { ToastContainer } from "react-toastify";
import "react-toastify/dist/ReactToastify.css";
import "./App.css";

function App() {
  const [contacts, handleContacts] = useState([]);
  const [showModal, handleShowModal] = useState(false);
  const [title, handleTitle] = useState("Contact Information");
  const [modalType, handleModalType] = useState("");
  const [selectedID, setSelectedID] = useState("");
  const [deleteBtn, setDeleteBtn] = useState(false);
  const [detail, setDetail] = useState([]);

  const CONSTANTS = {
    contact: "1",
    viewdetails: "2",
    editdetails: "3",
    deletedetails: "4",
  };

  const addContacts = (data) => {
    handleShowModal(false);
    const arr = [...contacts, data];
    handleContacts(arr);
  };

  const handleViewDetails = (id) => {
    const detail = contacts.filter((contact) => contact.id === id)[0];
    setDetail(detail);
    handleModalType(CONSTANTS.viewdetails);
    handleShowModal(true);
    setSelectedID(id);
  };

  const handleEditContact = () => {
    handleModalType(CONSTANTS.editdetails);
    handleTitle("Edit Contact");
  };

  const saveEditedContact = (data) => {
    const contactList = contacts.filter((contact) => contact.id !== selectedID);
    const newList = [...contactList, data];
    handleContacts(newList);
    handleShowModal(false);
  };

  const handleCheckboxClick = (id) => {
    let arr = [...contacts];
    let checked = false;
    arr.forEach((contact) => {
      if (contact.id === id) contact.check = !contact.check;
      if (contact.check) checked = true;
    });
    handleContacts(arr);
    setDeleteBtn(checked);
  };

  const getCheckedContacts = () => {
    let total = 0;
    contacts.forEach((contact) => {
      if (contact.check) total += 1;
    });
    return total;
  };

  const handleDeleteContact = () => {
    const newarr = contacts.filter((contact) => !contact.check);
    handleContacts(newarr);
    handleShowModal(false);
    setDeleteBtn(false);
  };

  return (
    <>
      <ToastContainer />
      {showModal && (
        <Modal
          show={showModal}
          handleClose={() => handleShowModal(false)}
          title={deleteBtn ? "Delete Contact(s)" : title}
          body={
            modalType === CONSTANTS.contact ||
            modalType === CONSTANTS.editdetails ? (
              <AddContact
                addContacts={addContacts}
                enableEditing={modalType === CONSTANTS.editdetails}
                saveEditedContact={saveEditedContact}
                details={detail}
              />
            ) : modalType === CONSTANTS.viewdetails ? (
              <ContactDetails
                details={detail}
                handleEditContact={handleEditContact}
              />
            ) : modalType === CONSTANTS.deletedetails ? (
              <div className="text-center m-5">
                You are about to delete {getCheckedContacts()} contact(s).
              </div>
            ) : null
          }
          footer={
            deleteBtn ? (
              <button className="btn btn-danger" onClick={handleDeleteContact}>
                CONFIRM
              </button>
            ) : null
          }
        />
      )}
      <div className="container p-5">
        <div className="display-end mt-4">
          <button
            className={`btn ${deleteBtn ? "btn-danger" : "btn-primary"}`}
            onClick={() => {
              handleShowModal(true);
              handleModalType(
                deleteBtn ? CONSTANTS.deletedetails : CONSTANTS.contact
              );
            }}
          >
            {deleteBtn ? "Delete Checked Contact(s)" : "Add Contact"}
          </button>
        </div>
        <div className="mt-5">
          {contacts.length > 0 ? (
            <ContactList
              arr={contacts}
              handleViewDetails={handleViewDetails}
              handleCheckboxClick={handleCheckboxClick}
            />
          ) : (
            <div className="text-center pt-5">You have no contacts!</div>
          )}
        </div>
      </div>
    </>
  );
}

export default App;
