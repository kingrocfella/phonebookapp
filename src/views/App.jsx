import { useState } from "react";
import Modal from "../components/Modal";
import AddContact from "./addContact";
import ContactList from "./contactList";
import ContactDetails from "./contactDetails";
import { ToastContainer, toast } from "react-toastify";
import "react-toastify/dist/ReactToastify.css";
import "../styles/App.scss";
import {
  BsPersonPlus,
  BsPersonDash,
  BsExclamationCircle,
} from "react-icons/bs";

function App() {
  const [contacts, handleContacts] = useState([]);
  const [showModal, handleShowModal] = useState(false);
  const [title, handleTitle] = useState("Contact Information");
  const [modalType, handleModalType] = useState("");
  const [selectedID, setSelectedID] = useState("");
  const [deleteBtn, setDeleteBtn] = useState(false);
  const [detail, setDetail] = useState({});

  const CONSTANTS = {
    contact: "1",
    viewdetails: "2",
    editdetails: "3",
    deletedetails: "4",
  };

  const addContacts = (data) => {
    handleShowModal(false);
    const arr = [...contacts, data];
    handleContacts(arr);
    toast.success("Contact created!", {
      position: toast.POSITION.TOP_RIGHT,
    });
  };

  const handleViewDetails = (id) => {
    const detail = contacts.filter((contact) => contact.id === id)[0];
    setDetail(detail);
    handleModalType(CONSTANTS.viewdetails);
    handleShowModal(true);
    setSelectedID(id);
  };

  const handleEditContact = () => {
    handleModalType(CONSTANTS.editdetails);
    handleTitle("Edit Contact");
  };

  const saveEditedContact = (data) => {
    const contactList = contacts.filter((contact) => contact.id !== selectedID);
    const newList = [...contactList, data];
    handleContacts(newList);
    handleShowModal(false);
    toast.success("Contact successfully edited!", {
      position: toast.POSITION.TOP_RIGHT,
    });
  };

  const handleCheckboxClick = (id) => {
    let arr = [...contacts];
    let checked = false;
    arr.forEach((contact) => {
      if (contact.id === id) contact.check = !contact.check;
      if (contact.check) checked = true;
    });
    handleContacts(arr);
    setDeleteBtn(checked);
  };

  const getCheckedContacts = () => {
    let total = 0;
    contacts.forEach((contact) => {
      if (contact.check) total += 1;
    });
    return total;
  };

  const handleDeleteContact = () => {
    const newarr = contacts.filter((contact) => !contact.check);
    handleContacts(newarr);
    handleShowModal(false);
    setDeleteBtn(false);
    toast.success("Contact(s) deleted!", {
      position: toast.POSITION.TOP_RIGHT,
    });
  };

  return (
    <>
      <ToastContainer />
      {showModal && (
        <Modal
          show={showModal}
          handleClose={() => handleShowModal(false)}
          title={
            deleteBtn
              ? "Delete Contact(s)"
              : modalType === CONSTANTS.contact
              ? "Add Contact"
              : title
          }
          body={
            modalType === CONSTANTS.contact ||
            modalType === CONSTANTS.editdetails ? (
              <AddContact
                addContacts={addContacts}
                enableEditing={modalType === CONSTANTS.editdetails}
                saveEditedContact={saveEditedContact}
                details={detail}
              />
            ) : modalType === CONSTANTS.viewdetails ? (
              <ContactDetails
                details={detail}
                handleEditContact={handleEditContact}
              />
            ) : modalType === CONSTANTS.deletedetails ? (
              <div className="text-center m-5">
                You are about to delete {getCheckedContacts()} contact(s).
              </div>
            ) : null
          }
          footer={
            deleteBtn ? (
              <button
                className="btn delete-button"
                onClick={handleDeleteContact}
              >
                CONFIRM
              </button>
            ) : null
          }
        />
      )}
      <div className="container text-white">
        <div className="space-between mt-5">
          <h3>Phonebook App</h3>
          <button
            className={`btn datatest-2 ${deleteBtn ? "delete-button" : "primary-button"}`}
            onClick={() => {
              handleShowModal(true);
              handleModalType(
                deleteBtn ? CONSTANTS.deletedetails : CONSTANTS.contact
              );
            }}
          >
            {deleteBtn ? <BsPersonDash /> : <BsPersonPlus />} &nbsp;&nbsp;
            {deleteBtn ? "Delete Checked Contact(s)" : "Add Contact"}
          </button>
        </div>
        <div className="mt-5">
          {contacts.length > 0 ? (
            <ContactList
              arr={contacts}
              handleViewDetails={handleViewDetails}
              handleCheckboxClick={handleCheckboxClick}
            />
          ) : (
            <div className="text-center pt-5 datatest-1">
              <BsExclamationCircle /> &nbsp; You have no contacts!
            </div>
          )}
        </div>
      </div>
    </>
  );
}

export default App;
