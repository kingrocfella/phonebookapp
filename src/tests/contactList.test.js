import React from "react";
import { mount } from "enzyme";
import App from "../views/contactList";

let container;
let props;
let mockCallBack;
let table;
let row;
describe("<App />", () => {
  beforeEach(() => {
    mockCallBack = jest.fn();
    props = {
      arr: [
        { firstName: "test", lastName: "test", id: 1, check: false },
        { firstName: "test", lastName: "test", id: 2, check: false },
        { firstName: "test", lastName: "test", id: 3, check: false },
      ],
      handleViewDetails: mockCallBack,
      handleCheckboxClick: mockCallBack,
    };
    container = mount(<App {...props} />);
    table = container.find("table");
    row = container.find("tr");
  });

  it("should match the snapshot", () => {
    expect(container.html()).toMatchSnapshot();
  });

  it("render the table and the rows depending on the contact list length", () => {
    expect(table.length).toBe(1);
    expect(row.length).toBe(props.arr.length);
  });

  it("Render checkbox for each row on the table", () => {
    expect(container.find('input[type="checkbox"]').length).toBe(
      props.arr.length
    );
  });

  it("handleCheckboxClick func should be called once checkbox is clicked", () => {
    container
      .find('input[type="checkbox"]')
      .at(1)
      .simulate("change", { target: { checked: true } });
    expect(mockCallBack.mock.calls.length).toEqual(1);
  });

  it("view details button for a row should be visible once that row is hovered over", () => {
    const btnRow = row.at(1).simulate("mouseenter");
    expect(btnRow.find("button").length).toEqual(1);
  });

  it("handleViewDetails func should be called once the view details button for a row is clicked", () => {
    const btnRow = row.at(1).simulate("mouseenter");
    btnRow.find("button").simulate("click");
    expect(mockCallBack.mock.calls.length).toEqual(1);
  });
});
