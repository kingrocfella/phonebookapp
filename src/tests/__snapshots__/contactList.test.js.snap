// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should match the snapshot 1`] = `"<table class=\\"table table-borderless contact-table text-white pt-5\\"><tbody><tr class=\\"cursor \\"><td class=\\"display-center-end pt-4\\"><div class=\\"check-container\\"><input type=\\"checkbox\\" id=\\"1\\"><span class=\\"checkmark\\"></span></div></td><td><div class=\\"display-vertical-center\\"><div class=\\"user-avatar display-center\\">TT</div><div class=\\"display-vertical-center pl-4\\">test test</div></div></td><td><div class=\\"display-end\\"><button class=\\"btn primary-button invisible \\" type=\\"button\\"><svg stroke=\\"currentColor\\" fill=\\"currentColor\\" stroke-width=\\"0\\" viewBox=\\"0 0 16 16\\" height=\\"1em\\" width=\\"1em\\" xmlns=\\"http://www.w3.org/2000/svg\\"><path d=\\"M10.5 8a2.5 2.5 0 11-5 0 2.5 2.5 0 015 0z\\"></path><path fill-rule=\\"evenodd\\" d=\\"M0 8s3-5.5 8-5.5S16 8 16 8s-3 5.5-8 5.5S0 8 0 8zm8 3.5a3.5 3.5 0 100-7 3.5 3.5 0 000 7z\\" clip-rule=\\"evenodd\\"></path></svg> &nbsp; View Details</button></div></td></tr><tr class=\\"cursor \\"><td class=\\"display-center-end pt-4\\"><div class=\\"check-container\\"><input type=\\"checkbox\\" id=\\"2\\"><span class=\\"checkmark\\"></span></div></td><td><div class=\\"display-vertical-center\\"><div class=\\"user-avatar display-center\\">TT</div><div class=\\"display-vertical-center pl-4\\">test test</div></div></td><td><div class=\\"display-end\\"><button class=\\"btn primary-button invisible \\" type=\\"button\\"><svg stroke=\\"currentColor\\" fill=\\"currentColor\\" stroke-width=\\"0\\" viewBox=\\"0 0 16 16\\" height=\\"1em\\" width=\\"1em\\" xmlns=\\"http://www.w3.org/2000/svg\\"><path d=\\"M10.5 8a2.5 2.5 0 11-5 0 2.5 2.5 0 015 0z\\"></path><path fill-rule=\\"evenodd\\" d=\\"M0 8s3-5.5 8-5.5S16 8 16 8s-3 5.5-8 5.5S0 8 0 8zm8 3.5a3.5 3.5 0 100-7 3.5 3.5 0 000 7z\\" clip-rule=\\"evenodd\\"></path></svg> &nbsp; View Details</button></div></td></tr><tr class=\\"cursor \\"><td class=\\"display-center-end pt-4\\"><div class=\\"check-container\\"><input type=\\"checkbox\\" id=\\"3\\"><span class=\\"checkmark\\"></span></div></td><td><div class=\\"display-vertical-center\\"><div class=\\"user-avatar display-center\\">TT</div><div class=\\"display-vertical-center pl-4\\">test test</div></div></td><td><div class=\\"display-end\\"><button class=\\"btn primary-button invisible \\" type=\\"button\\"><svg stroke=\\"currentColor\\" fill=\\"currentColor\\" stroke-width=\\"0\\" viewBox=\\"0 0 16 16\\" height=\\"1em\\" width=\\"1em\\" xmlns=\\"http://www.w3.org/2000/svg\\"><path d=\\"M10.5 8a2.5 2.5 0 11-5 0 2.5 2.5 0 015 0z\\"></path><path fill-rule=\\"evenodd\\" d=\\"M0 8s3-5.5 8-5.5S16 8 16 8s-3 5.5-8 5.5S0 8 0 8zm8 3.5a3.5 3.5 0 100-7 3.5 3.5 0 000 7z\\" clip-rule=\\"evenodd\\"></path></svg> &nbsp; View Details</button></div></td></tr></tbody></table>"`;
